Migrate the ImportAdd task to coroutine

lukstbit opened the pull request:
## Purpose / Description

Part of https://github.com/ankidroid/Anki-Android/issues/7108. The minimum amount of work to migrate this task to coroutine to remove it from CollectionTask. There's no need for extra refactoring or TODOs as the changed code is on the legacy pathway and will be removed when the new backend becomes the default..

## How Has This Been Tested?

Ran the tests, manually verified the import feature.

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

Arthur-Milchior commented:
@lukstbit you've got a conflict

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1890,14 +1892,27 @@ open class DeckPicker :
                 Timber.d("doInBackgroundImportAdd")
                 @Suppress("DEPRECATION")
                 progressDialog.setMessage(resources.getString(R.string.import_title))
+                val progressCallback = kotlinx.coroutines.channels.Channel<ImportAddProgress>()
```
Thanks for the commit message. I find this code hard to read.
Part of it may be that I'm really not used to this way of writting code, I'll admit that. Still, I suspect that we should have some comment in the code.
The easiest path would probably be to have a function that create the channel, the documentation of the function can then everything there is to know.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1906,7 +1921,7 @@ open class DeckPicker :
                     }
                 }
-
+                progressCallback.close()
```
I'm really not confortable here. A lot of things occurs. Many of them have plenty of reason to throw exception. We should theoretically catch them. Still, I'd strongly prefer that we would transform this channel into a Closeable, and put all the code if y `.use` to be certain we actually close it. The extra function mentioned above could help with that too

lukstbit replied:
You are right. I've added a new helper method in CoroutineHelpers to fix this. I also added documentation to that method to explain about its usage, hopefully that's more clear.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1890,14 +1892,27 @@ open class DeckPicker :
                     val imp = AnkiPackageImporter(col, path)
-                    // needs to execute the previous onProgress callback code: context.mProgressDialog!!.setMessage(value)
-                    imp.setProgressCallback(null)
+                    imp.setProgressCallback(progressCallback)
```
I never noticed before, but there is a problem here. We are considering importing many files. However, the message only state the percentage of the current file. We should ensure that, one way or another, if we import multiple files, the message should state which file is currently being imported.
Since it would be a functional change, I'd be fine with a TODO plus an issue, if you so please

lukstbit replied:
I've added a TODO where the progress message is set on the dialog to mention about showing the name of the imported file. The filename is available and can be passed in, but the problem in fixing this is going to be displaying it. We are combining all progress data into a single string and this makes it difficult to properly handle the filename(like when it's too big for example).

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.kt:
```diff
@@ -877,3 +883,9 @@ open class Anki2Importer(col: Collection?, file: String) : Importer(col!!, file)
+data class ImportAddProgress(
+    val notesDone: Int,
+    val cardsDone: Int,
+    val postProcess: Int,
```
Would you please mind documenting. `postProcess` seems really not clear to me. The other ones are, I guess, okay.

lukstbit replied:
I added the original documentation from the publishProgress() method.

Arthur-Milchior reviewed (commented):
Thanks a lot. 
I believe last commit could be made safer. I may also be wrong and would not oppose if another reviewer feel confortable approving it

mikehardy commented:
Conflict again 😞 - but we'll get this in I'm sure - sorry + thanks

BrayanDSO commented:
Bumping priority since this may help solve an issue

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1607,12 +1610,102 @@ open class DeckPicker :
+    }
+
+    private fun importAddPostTask(result: ImporterData?) {
+        if (mProgressDialog != null && mProgressDialog!!.isShowing) {
```
```
mProgressDialog?.apply {
   if(isShowing) dismiss
}
```

The risk it actually becomes null in the meantime is probably very small if it exists at all. Still, slight improvement

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1607,12 +1610,102 @@ open class DeckPicker :
+        }
+    }
+
+    private fun importAddPostTask(result: ImporterData?) {
```
Let's remove the `?` since the first thing you do is use `!!` on the result.

Arthur-Milchior reviewed (approved):
Thanks @lukstbit 
Feel free to push back on my nits. Everything good otherwise.

For the sake of simplifying the reviews, I loved the way changes were done last summer, with multiple commits with smaller changes, so that I can easily check what is moved where, if that's not to hard to do, I'd appreciate for a next time. But if it's too time consuming, never mind

lukstbit commented:
> For the sake of simplifying the reviews, I loved the way changes were done last summer, with multiple commits with smaller changes,

This was the way this PR was initially done but when I came back to this it required me to fix conflicts for each of the commits so I simply squash. But you're right and I'll use several commits for other tasks refactoring.

I addressed the last issues, also fixed a warning issued(correctly) by the IDE.
